### PR TITLE
better to_tree function

### PR DESCRIPTION
**Pull request by awalterschulze**
A better to_tree function as proposed by @Nielius on the mailing list.
Proofs still seem to work and I think this is a much better definition.

### CODE REVIEW

**Comment by awalterschulze** [comment]
Although maybe
```
Fixpoint to_tree (xs: list X) : tree X :=
    match xs with
    | nil => value Identity
    | (x'::xs') => bin (value x') (to_tree xs')
    end.
```
Is an even more logical definition.

**Comment by awalterschulze** [comment]
But then the current proofs start to break.

**Comment by Nielius** [comment]
Yeah, all three definitions are slightly different, but I think the last one you suggested may be the best. I'm pretty sure we can still prove everything using that definition, if we wanted to.

Your last suggestion is pretty good, I think, because it is the most uniform: every tree will look like
```
(bin (val x_1) (bin (val x_2) (bin (val x_3) ..... (bin (val x_n) (val Id)))))
=
   .
  / \
x_1  \
      .
     / \
   x_2   \
          .
         / \
       x_n  id
```
whereas the other two will always end in
```
(bin (val x_{n-1}) (val x_n)) =
          .
         / \
    x_n-1   x_n
```

But then the difference between the original definition and the one in this PR (my suggestion), is that the original one maps `(x::nil)` to `(bin (val x) (val nil))`, whereas mine maps it to just `(val x)`.

So in short, they are all slightly different, I think your last one is cleanest, and I also think we can prove every statement we want with any of these definitions.

**Comment by awalterschulze** [comment]
Maybe we could merge this and then create another to_tree function, which we can also use in new proofs, so we don't break the current proofs and can also compare with our old proofs.

What do you think?

**Comment by Nielius** [comment]
Yeah, that seems like a good idea. I'll merge this for now.
